Mathlib/Order/ReltoGaloisConnection.lean

```diff
@@ -0,0 +1,114 @@
+/-
+Copyright (c) 2024 Lagrange Mathematics and Computing Research Center. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anthony Bordg
+-/
+import Mathlib.Data.Rel
+import Mathlib.Order.GaloisConnection
+
+/-!
+# The Galois Connection Induced by a Relation
+
+In this file, we show that an arbitrary relation `R` between a pair of types `α` and `β` defines
+a pair `leftDual` and `rightDual` of adjoint order-preserving maps between the corresponding posets
+`Set α` and `(Set β)ᵒᵈ`.
+We define `lFixedPoints` (resp. `rFixedPoints`) as the set of fixed points `J` (resp. `I`) of
+`Set α` (resp. `(Set β)ᵒᵈ`) such that `rightDual (leftDual J) = J` (resp.
+`leftDual (rightDual I) = I`).
+
+## Main Results
+
+⋆ `galoisConnection`: we prove that the maps `leftDual` and `rightDual` form a Galois connection.
+⋆ `equivFixedPoints`: we prove that the said maps induce inverse bijections between the sets
+of fixed points.
+
+## References
+
+⋆ Engendrement de topologies, démontrabilité et opérations sur les sous-topos, Olivia Caramello and
+Laurent Lafforgue (in preparation)
+
+## Tags
+
+relation, Galois connection, induced bijection, fixed points
+-/
+
+section
+
+universe u v
+
+variable {α : Type u} {β : Type v} (R : Rel α β)
+
+namespace Rel
+
+/-! ### Pairs of adjoint maps defined by relations -/
+
+open OrderDual
+
+/-- `leftDual` maps any set `J` of elements of type `α` to the set `{b : β | ∀ a ∈ J, R a b}` of
+elements `b` of type `β` such that `R a b` for every element `a` of `J`. -/
+def leftDual (J : Set α) : (Set β)ᵒᵈ := toDual {b : β | ∀ ⦃a⦄, a ∈ J → R a b}
+
+/-- `rightDual` maps any set `I` of elements of type `β` to the set `{a : α | ∀ b ∈ ofDual I, R a b}`
+of elements `a` of type `α` such that `R a b` for every element `b` of `I`. -/
+def rightDual (I : (Set β)ᵒᵈ) : Set α := {a : α | ∀ ⦃b⦄, b ∈ ofDual I → R a b}
+
+/-- The pair of functions `leftDual` and `rightDual` forms a Galois connection. -/
+theorem galoisConnection : GaloisConnection (leftDual R) (rightDual R) := by
```

Comment:
... and
```suggestion
theorem galoisConnection : GaloisConnection (toDual ∘ leftDual R) (rightDual R ∘ ofDual) := by
```

Reply:
@YaelDillies @adamtopaz 
I don't see what you really gain with these changes, however please note that the definition of `rFixedPoints` will be far more cumbersome, in particular you will break the symmetry in the expressions of `lFixedPoints` and `rFixedPoints`. Moreover, you won't be able to remove the mention of the dual order `(Set β)ᵒᵈ` in `rFixedPoints`. 
Is it really what you want?

Reply:
Why can't it simply be
```
def rFixedPoints := {I : Set β | R.leftDual (R.rightDual I) = I}
```
?

Reply:
In this case, how will you prove `leftDual_mem_rFixedPoint`, i.e. that an image is a fixed-point?

Reply:
Same proof. Why would it change? You're just adding a bunch of "identity" isomorphisms here and there, and they cancel out

Reply:
The adjoint maps now are (`toDual \circ leftDual`, `rightDual \circ ofDual`), not (`leftDual`, `rightDual`) anymore. You can still express the property of `J` being a left fixed point with `R.rightDual (R.leftDual J) = J`, since `(rightDual \circ ofDual) \circ (toDual \circ leftDual) = rightDual \circ leftDual`, i.e. the additional isos cancel out, but for a right fixed point they don't cancel out, the property of being a right fixed point have to use `(toDual \circ leftDual) \circ (rightDual \circ ofDual)` \neq `\leftDual \circ rightDual` (these last two maps are not even of the same type, i.e. automorphism of `(Set β)ᵒᵈ` vs of `Set β`).
Am I missing something?

Reply:
Sorry, not at a computer right now but morally if your statements are not about orders, then they shouldn't need anything related to `OrderDual`. `rFixedPoints_mem_rDual` isn't about order so you shouldn't need `OrderDual`, even though the intermediate lemma `galoisConnection` needs it.

Reply:
Thank you for your reply and sorry to bother on a Sunday, Yael :)
>if your statements are not about orders

I believe there are, no? $I$ being a fixed point under $F$ means $I \leq F (I)$ and $F(I)\leq I$.
Moreover, if these statements weren't about orders, you would prevent yourself from using the API for Galois connections ... so, there would be no point in proving that the induced bijections between fixed points come from a Galois connection. 
Since the change was suggested by Adam, maybe he can help?

Reply:
I'm not saying it *can't* be interpreted as an order theoretic statement, but rather that it *can* be interpreted without order theory. If all else fails, I'll be back at a computer first thing tomorrow morning ;)

Reply:
Done. Does our suggestion make more sense now?

Reply:
Thank you, Yael. I left a comment below https://github.com/leanprover-community/mathlib4/pull/14908#discussion_r1686241361